hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2012-2016 JetBrains s.r.o
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package jetbrains.jetpad.hybrid;
+
+import jetbrains.jetpad.base.Handler;
+import jetbrains.jetpad.base.Registration;
+import jetbrains.jetpad.hybrid.parser.Parser;
+import jetbrains.jetpad.hybrid.parser.ParsingContextFactory;
+import jetbrains.jetpad.hybrid.parser.Token;
+import jetbrains.jetpad.hybrid.parser.prettyprint.PrettyPrinter;
+import jetbrains.jetpad.hybrid.parser.prettyprint.PrettyPrinterContext;
+import jetbrains.jetpad.model.collections.CollectionItemEvent;
+import jetbrains.jetpad.model.collections.CollectionListener;
+import jetbrains.jetpad.model.collections.list.ObservableArrayList;
+import jetbrains.jetpad.model.collections.list.ObservableList;
+import jetbrains.jetpad.model.event.EventHandler;
+import jetbrains.jetpad.model.event.ListenerCaller;
+import jetbrains.jetpad.model.event.Listeners;
+import jetbrains.jetpad.model.property.PropertyChangeEvent;
+
+import java.util.List;
+import java.util.Objects;
+
+public class PrettyHybridProperty<SourceT, ModelT> implements HybridProperty<ModelT> {
+  private final ObservableList<Token> myTokens = new MyList();
+  private ModelT myValue;
+
+  private final ObservableList<SourceT> mySource;
+  private final IndexedTransform<Token, SourceT> myFrom;
+  private final Handler<Integer> myRemoveHandler;
+  private final Handler<Runnable> myUpdateHandler;
+  private final Parser<? extends ModelT> myParser;
+  private final PrettyPrinter<? super ModelT> myPrinter;
+  private final ParsingContextFactory myParsingContextFactory;
+
+  private boolean myInSync = false;
+  private Listeners<EventHandler<? super PropertyChangeEvent<ModelT>>> myHandlers;
+
+  public PrettyHybridProperty(
+      ObservableList<SourceT> source,
+      final IndexedTransform<SourceT, Token> to,
+      IndexedTransform<Token, SourceT> from,
+      Handler<Integer> removeHandler,
+      Handler<Runnable> updateHandler,
```

Comment:
Remove handler need to communicate token removal to application. Update handler needed to wrap source operation, for instance into OT transaction.